src/julia-syntax.scm

```diff
@@ -2283,17 +2283,7 @@
 
    'typed_comprehension
    (lambda (e)
-     (expand-forms
-      (or (and (eq? (caaddr e) 'generator)
-               (let ((ranges (cddr (caddr e))))
-                 (if (any (lambda (x) (eq? x ':)) ranges)
-                     (error "comprehension syntax with `:` ranges has been removed"))
-                 (and (every (lambda (x) (and (pair? x) (eq? (car x) '=)))
-                             ranges)
-                      ;; TODO: this is a hack to lower simple comprehensions to loops very
-                      ;; early, to greatly reduce the # of functions and load on the compiler
```

Comment:
See this comment.

Reply:
yes. and to expand on that a little, we use this to make it possible to use comprehensions in reflection and inference. we wouldn't be able to use comprehensions there if it wasn't for this guarantee.

Reply:
OK, I guess that makes sense; I just noticed I can even return a typed comprehension from a `@generated` function (for obvious reasons). There seems to be a bunch of `Any[... for ...]` comprehensions in the compiler.

On the other hand, it seems these days that `collect` is getting some overloads and therefore untyped comprehensions *aren't* so tied to `Array` anymore. For example, is it bad that `[f(x) for x in a::StaticArray] isa StaticArray`? This is juxtaposed with `!(Float64[f(x) for x in a::StaticArray) isa StaticArray)`. I'm wondering if we can get a bit more consistency? (As in - maybe the solution we want is the opposite of this PR, and enforce the untyped version to always create an `Array`?)

I note that I don't see an aweful lot of typed comprehensions in user code (maybe I'm wrong and that's just me?) so do you know if outside of Julia internals does this lowering "optimization" have much of an impact on latency in the wild? If not, could the desired behavior be viewed as more of a bootstrap vs not bootstrap issue?

Reply:
Yes I definitely read this comment! At face value it says "this is a performance hack" and I wondered whether it was still necessary. Especially because comprehensions seem more common than typed comprehensions for user code.

Thanks @vtjnash for expanding. Here's what I think you're saying:
* Normal comprehensions are prohibited in `base/compiler` due to the call to `return_type` here: https://github.com/JuliaLang/julia/blob/442d1599e62b1a3a453bd0c28c9a795e09e939ff/base/array.jl#L594
* Simple typed comprehensions are allowed because they lower to a simple form which doesn't re-enter inference.
* The comment `greatly reduce the # of functions and load on the compiler` refers to the performance of comprehensions in the compiler itself rather than the performance of user code (??)

But I don't understand why lowering to `collect(T, gen)` is not ok - this doesn't call `return_type`; it goes via the `_collect` methods here: https://github.com/JuliaLang/julia/blob/442d1599e62b1a3a453bd0c28c9a795e09e939ff/base/array.jl#L519

If this is about compiler performance, what should I be measuring? Is the time to compile Base relevant?